stdlib/public/core/StringIndex.swift

```diff
@@ -9,133 +9,266 @@
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 //===----------------------------------------------------------------------===//
-extension String {
-  /// A position of a character or code unit in a string.
-  @_fixed_layout // FIXME(sil-serialize-all)
-  public struct Index {
-    @usableFromInline
-    internal typealias _UTF8Buffer = UTF8.EncodedScalar
 
-    @usableFromInline // FIXME(sil-serialize-all)
-    internal var _compoundOffset: UInt64
+import SwiftShims
 
-    @usableFromInline
-    internal var _utf8Buffer = _UTF8Buffer()
+/*
 
-    @usableFromInline
-    internal var _graphemeStrideCache: UInt16 = 0
-  }
-}
+String's Index has the following layout:
+
+ ┌──────────┬───────────────────┬────────────────┬──────────┐
+ │ b63:b16  │      b15:b14      │     b13:b8     │ b7:b0    │
+ ├──────────┼───────────────────┼────────────────┼──────────┤
+ │ position │ transcoded offset │ grapheme cache │ reserved │
+ └──────────┴───────────────────┴────────────────┴──────────┘
+
+- grapheme cache: A 6-bit value remembering the distance to the next grapheme
+boundary
+- position aka `encodedOffset`: An offset into the string's code units
+- transcoded offset: a sub-scalar offset, derived from transcoding
+
+The use and interpretation of both `reserved` and `grapheme cache` is not part
+of Index's ABI; it should be hidden behind non-inlinable calls. However, the
+position of the sequence of 14 bits allocated is part of Index's ABI, as well as
+the default value being `0`.
+
+*/
 
-/// Convenience accessors
 extension String.Index {
-  @inlinable // FIXME(sil-serialize-all)
-  internal var utf8Buffer: String.Index._UTF8Buffer? {
-    guard !_utf8Buffer.isEmpty else { return nil }
-    return _utf8Buffer
+  @inlinable
+  internal var orderingValue: UInt64 {
+    @inline(__always) get { return _rawBits &>> 14 }
   }
 
-  @inlinable // FIXME(sil-serialize-all)
+  // Whether this is at the canonical "start" position, that is encoded AND
+  // transcoded offset of 0.
+  @inlinable
+  internal var isZeroPosition: Bool {
+    @inline(__always) get { return orderingValue == 0 }
+  }
+
+  /// The offset into a string's code units for this index.
+  @inlinable
+  public var encodedOffset: Int {
+    @inline(__always) get { return Int(truncatingIfNeeded: _rawBits &>> 16) }
+  }
+
+  @inlinable
+  internal var transcodedOffset: Int {
+    @inline(__always) get {
+      return Int(truncatingIfNeeded: orderingValue & 0x3)
+    }
+  }
+
+  @usableFromInline
   internal var characterStride: Int? {
-    guard _graphemeStrideCache > 0 else { return nil }
-    return Int(truncatingIfNeeded: _graphemeStrideCache)
+    let value = (_rawBits & 0x00FC_0000_0000_0000) &>> 50
```

Comment:
This doesn't match the position of the grapheme cache in the layout description above!

Reply:
Huh, that might be why grapheme breaking is 2x slower in some ad-hoc tests I did. It's almost as if it's measuring the distance twice because nothing's in the cache... 😅

Good catch!